src/backend/distributed/commands/foreign_constraint.c

```diff
@@ -149,237 +150,291 @@ ErrorIfUnsupportedForeignConstraint(Relation relation, char distributionMethod,
-						  's', &referencedColumnArray, NULL, &referencedColumnCount);
+		heapTuple = systable_getnext(scanDescriptor);
+	}
 
-		Assert(referencingColumnCount == referencedColumnCount);
+	/* clean up scan and close system catalog */
+	systable_endscan(scanDescriptor);
+	heap_close(pgConstraint, AccessShareLock);
+}
 
-		for (attrIdx = 0; attrIdx < referencingColumnCount; ++attrIdx)
-		{
-			AttrNumber referencingAttrNo = DatumGetInt16(referencingColumnArray[attrIdx]);
-			AttrNumber referencedAttrNo = DatumGetInt16(referencedColumnArray[attrIdx]);
-
-			if (distributionColumn->varattno == referencingAttrNo &&
-				(!referencedTableIsAReferenceTable &&
-				 referencedTablePartitionColumn->varattno == referencedAttrNo))
-			{
-				foreignConstraintOnPartitionColumn = true;
-			}
-
-			if (distributionColumn->varattno == referencingAttrNo)
-			{
-				referencingColumnsIncludeDistKey = true;
-			}
-		}
 
+/*
+ * Errors if we don't support a foreign key constraint from the given referencing
+ * table to the given referenced table.
+ *
+ * To support foreign constraints, we require that;
+ * - If referencing and referenced tables are hash-distributed
+ *		- Referencing and referenced tables are co-located.
+ *      - Foreign constraint is defined over distribution column.
+ *		- ON DELETE/UPDATE SET NULL, ON DELETE/UPDATE SET DEFAULT and ON UPDATE CASCADE options
+ *        are not used.
+ *      - Replication factors of referencing and referenced table are 1.
+ * - If referenced table is a reference table
+ *      - ON DELETE/UPDATE SET NULL, ON DELETE/UPDATE SET DEFAULT and ON UPDATE CASCADE options
+ *        are not used on the distribution key of the referencing column.
+ * - If referencing table is a reference table, error out if the referenced table is not a
+ *   a reference table.
+ */
+static void
+ErrorIfUnsupportedForeignConstraint(char referencingDistMethod,
+									Var *referencingDistColumn,
+									uint32 referencingColocationId,
+									bool referencingSingleReplicated,
+									char referencedDistMethod,
+									Var *referencedDistColumn,
+									uint32 referencedColocationId,
+									HeapTuple pgConstraintTuple)
+{
+	bool referencingIsReferenceTable = (referencingDistMethod == DISTRIBUTE_BY_NONE);
+	bool referencedIsReferenceTable = (referencedDistMethod == DISTRIBUTE_BY_NONE);
+	Form_pg_constraint constraintForm = (Form_pg_constraint) GETSTRUCT(pgConstraintTuple);
+	int referencingAttrIndex = -1;
+	int referencedAttrIndex = -1;
+	bool referencingColumnsIncludeDistKey = false;
+	bool foreignConstraintOnDistKey = false;
 
-		/*
-		 * If columns in the foreign key includes the distribution key from the
-		 * referencing side, we do not allow update/delete operations through
-		 * foreign key constraints (e.g. ... ON UPDATE SET NULL)
-		 */
+	/*
+	 * We support foreign keys between reference tables. No more checks
+	 * are necessary.
+	 */
+	if (referencingIsReferenceTable && referencedIsReferenceTable)
```

Comment:
Except this line + removal of the related error, everything else in this file is just moving things around and renaming variables so code can be understood a bit easier.